pkg/storage/db.go

```diff
@@ -0,0 +1,278 @@
+package storage
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"log"
+	"sync"
+	"time"
+
+	"github.com/asecurityteam/asset-inventory-api/pkg/domain"
+	"github.com/asecurityteam/asset-inventory-api/pkg/logs"
+
+	_ "github.com/lib/pq" // postgres driver for sql must be imported so sql finds and uses it
+)
+
+const tableAWSResources = "aws_resources"
+const tableAWSIPS = "aws_ips"
+const tableAWSHostnames = "aws_hostnames"
+const tableAWSEventsIPSHostnames = "aws_events_ips_hostnames"
+
+// can't use Sprintf in a const, so...
+const commonQuery = "SELECT " +
+	"	aws_resources_id, aws_ips_ip, aws_hostnames_hostname, is_public, is_join, ts, aws_resources.account_id, aws_resources.region, aws_resources.type, aws_resources.meta " +
+	"FROM " +
+	"	" + tableAWSEventsIPSHostnames + " " +
+	"JOIN " + tableAWSResources + " " +
+	"	ON aws_resources_id = aws_resources.id " +
+	"WHERE "
+
+// DB represents a convenient database abstraction layer
+type DB struct {
+	sqldb *sql.DB      // this is a unit test seam
+	LogFn domain.LogFn // also a seam
+
+	once sync.Once
+}
+
+// Init initializes a connection to a Postgres database according to the environment variables POSTGRES_HOST, POSTGRES_PORT, POSTGRES_USER, POSTGRES_PASSWORD, POSTGRES_DATABASE
+func (db *DB) Init(ctx context.Context, postgresConfig *PostgresConfig) error {
+	var initerr error
+	db.once.Do(func() {
+
+		host := postgresConfig.Hostname
+		port := postgresConfig.Port
+		user := postgresConfig.Username
+		password := postgresConfig.Password
+		dbname := postgresConfig.DatabaseName
+
+		logger := db.LogFn(ctx)
+
+		if db.sqldb == nil {
+			sslmode := "disable"
+			if host != "localhost" {
+				sslmode = "enable"
+			}
+			psqlInfo := fmt.Sprintf("host=%s port=%s user=%s "+
+				"password=%s dbname=%s sslmode=%s",
+				host, port, user, password, dbname, sslmode)
+			pgdb, err := sql.Open("postgres", psqlInfo)
+			if err != nil {
+				logger.Error(logs.DBOpenError{Reason: err.Error()})
+				initerr = err
+			}
+
+			err = pgdb.Ping()
+			if err != nil {
+				logger.Error(logs.DBPingError{Reason: err.Error()})
+				initerr = err
+			}
+
+			db.sqldb = pgdb
+		}
+	})
+	return initerr
+}
+
+// StoreCloudAsset an implementation of the Storage interface that records to a database
+func (db *DB) StoreCloudAsset(ctx context.Context, cloudAssetChanges domain.CloudAssetChanges) error {
+	tx, err := db.sqldb.Begin()
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		switch err {
+		case nil:
+			if commitErr := tx.Commit(); commitErr != nil {
+				log.Fatalf(`Failed to commit transation: %s`, commitErr)
```

Comment:
We should not log here, and we should almost never log at the Fatal level. Fatal will call `os.Exit()` and cause our app to terminate. This should be pulled out of the defer so that we can return these errors from the top level function.

Reply:
To extend this, we don't use the standard library logger anywhere because it does not enable use to use structured logging. Anything that would be logged should be logged using the logger from context and with a struct. This helps ensures that all of our logs are unique, contain the appropriate context, and can be marshalled to a structured output for log ingestion systems.

Reply:
I see.  Will fix.